### PR TITLE
Add variant system

### PR DESCRIPTION
**Pull request by collijk**
This is the ode code only for the new variant system corresponding to the attached diagram.  This is none of the wiring necessary to hook it up.  See later PRs for that.

image

### CODE REVIEW

File: src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py [hunk 1]
```diff
@@ -0,0 +1,581 @@
...
+            infectious_wild = infectious_wild + y[local_s + LOCAL_I1] + y[local_s + LOCAL_I2]
+
+        for j in range(N_VARIANT_TYPE_SEIIR_GROUPS):
+            local_s = i*GROUP_SYSTEM_SIZE + (j + N_WILD_TYPE_SEIIR_GROUPS)*N_SEIIR_COMPARTMENTS
+            infectious_variant = infectious_variant + y[local_s + LOCAL_I1] + y[local_s + LOCAL_I2]
+
+    infectious = np.array([infectious_wild, infectious_variant])
+
+    # Do the thing!
+    dy = np.zeros_like(y)
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py:117]: I likey zero's ;)

File: src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+
+s_m, r_ma = 35, 36
+
+# Composite tracking compartments.
+new_e_wild, new_e_variant, new_e_p_wild, new_e_p_variant = 37, 38, 39, 40
+new_e_nbt, new_e_vbt, new_s_v, new_r_w = 41, 42, 43, 44
+v_u, v_p, v_pa, v_m, v_ma = 45, 46, 47, 48, 49
+
+# Vaccine categories.
+u, p, pa, m, ma = 0, 1, 2, 3, 4
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py:67]: Suggestion: destructure COMPARTMENTS

```python
# massive destructuring - keeps us honest
(
    s, e, i1, i2, r ,
    s_u, e_u, i1_u, i2_u, r_u ,
    s_p, e_p, i1_p, i2_p, r_p ,
    s_pa, e_pa, i1_pa, i2_pa, r_pa ,

    s_variant, e_variant, i1_variant, i2_variant, r_variant ,
    s_u_variant, e_u_variant, i1_u_variant, i2_u_variant, r_u_variant ,
    s_pa_variant, e_pa_variant, i1_pa_variant, i2_pa_variant, r_pa_variant ,

    s_m, r_ma ,

    # Composite tracking compartments.
    new_e_wild, new_e_variant, new_e_p_wild, new_e_p_variant ,
    new_e_nbt, new_e_vbt, new_s_v, new_r_w ,
    v_u, v_p, v_pa, v_m, v_ma ,
) = range(len(COMPARTMENTS))

# proof I'm not betraying you
print(s)
print(s_m)
print(new_e_nbt)
print(v_ma)
```

```shell
$ python big_destructure.py 
0
35
41
49
```

File: src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py [hunk 3]
```diff
@@ -0,0 +1,581 @@
...
+        vaccines_out = get_unprotected_vaccines_from_not_s(
+            y, vaccines_out, params, n_unvaccinated, v_total,
+            e_variant, i1_variant, i2_variant, r_variant,
+        )
+
+    v_total_empirical = vaccines_out.sum()
+#    if v_total:
+#       percent_loss = abs(v_total_empirical - v_total) / v_total  * 100
+#       if percent_loss > .05:
+#           print('Losing more than 5% vaccines. Percent loss: ', percent_loss)
```
Comment (miker985) [inline src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py:434]: Is this dead code useful moving forward or should it be shed meow?

**Review by miker985 — APPROVED** [review]
I cannot comment on correctness beyond basic syntax.

My confidence level is high that you and at least 1 other person have some idea what this means.

File: src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py (hunk 2 above)
Comment (collijk) [reply src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py:67]: This is great.  I tried a few abstractions but they broke numba. This seems perfect.

File: src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py (hunk 3 above)
Comment (collijk) [reply src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py:434]: Still useful. It's the only place I have this system written down.

Comment (collijk) [reply src/covid_model_seiir_pipeline/pipeline/forecasting/model/ode_systems/variant.py:434]: Which is a bad use of code space, but I'm pressed for time.

**Pull request merged**